Dev -> Master for v3.2 release

drpatelh opened the pull request.

grst commented on README.md:
```diff
@@ -24,32 +24,33 @@ The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool
-5. Adapter and quality trimming ([`Trim Galore!`](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/))
-6. Removal of ribosomal RNA ([`SortMeRNA`](https://github.com/biocore/sortmerna))
-7. Choice of multiple alignment and quantification routes:
+The SRA download functionality has been removed from the pipeline (`>=3.2`) and ported to an independent workflow called [nf-core/fetchngs](https://nf-co.re/fetchngs). You can provide `--nf_core_pipeline rnaseq` when running nf-core/fetchngs to download and auto-create a samplesheet containing publicly available samples that can be accepted directly as input by this pipeline.
```
Strictly speaking, removing functionality would require a major release according to semver?

JoseEspinosa replied:
"Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backward compatible functionality is introduced to the public API. It MUST be incremented if any public API functionality is marked as deprecated." extracted from [semver](https://semver.org/). I think is OK just incrementing minor, isn't it?

drpatelh replied:
Yup, I think a minor version is ok here too.

grst commented on README.md:
````diff
@@ -69,30 +70,22 @@ The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool
-            --public_data_ids ids.txt \
-            -profile <docker/singularity/podman/conda/institute>
+        ```console
+        ~/.nextflow/assets/nf-core/rnaseq/bin/fastq_dir_to_samplesheet.py <FASTQ_DIR> samplesheet.csv
````
That would require a `nextflow pull` first. I would either add that, or use something like  `curl -s https://raw.githubusercontent.com/nf-core/rnaseq/dev/bin/fastq_dir_to_samplesheet.py | python <FASTQ_DIR> samplesheet.csv`

drpatelh replied:
Fixed https://github.com/nf-core/rnaseq/pull/660/commits/76dfe0c144fa6fe450739dad940b2f3f29176536

grst commented on bin/check_samplesheet.py:
```diff
@@ -67,10 +67,8 @@ def check_samplesheet(file_in, file_out):
-                if sample.find(" ") != -1:
-                    print_error("Sample entry contains spaces!", "Line", line)
-            else:
+            sample = sample.replace(' ', '_')
```
Although unlikely, this could, in theory, lead to name clashes.

drpatelh replied:
So you think we should keep the old behaviour and fail? I thought it would just be less hassle for users but you're right.

grst replied:
It's probably ok. How about writing out a warning message just in case, something like: 
```
NOTE: Spaces in sample <xxx> have been replaced by underscores
```

drpatelh replied:
Added https://github.com/nf-core/rnaseq/pull/660/commits/3eb2bac959b5987ae777325f232ee6b4c1a19791

grst commented on bin/fastq_dir_to_samplesheet.py:
```diff
@@ -0,0 +1,87 @@
+    parser.add_argument('-st', '--strandedness', type=str, dest="STRANDEDNESS", default='unstranded', help="Value for 'strandedness' in samplesheet. Must be one of 'unstranded', 'forward', 'reverse'.")
+    parser.add_argument('-r1', '--read1_extension', type=str, dest="READ1_EXTENSION", default='_R1_001.fastq.gz', help="File extension for read 1.")
+    parser.add_argument('-r2', '--read2_extension', type=str, dest="READ2_EXTENSION", default='_R2_001.fastq.gz', help="File extension for read 2.")
+    parser.add_argument('-se', '--single_end', dest="SINGLE_END", action='store_true', help="Single-end information will be auto-detected but this option forces paired-end FastQ files to be treated as single-end so only read 1 information is included in the samplesheet.")
```
these are super long lines that require me to scroll horizontally during code-review. Consider reformatting using `black`.

drpatelh replied:
Oops 👀 Ran `black` locally to fix the formatting https://github.com/nf-core/rnaseq/pull/660/commits/936d73a55850abe1197e35d79495028c7be2e0ea

And created an official issue on the tools repo to add this for linting CI.

grst commented on bin/fastq_dir_to_samplesheet.py:
```diff
@@ -0,0 +1,87 @@
+    parser.add_argument("FASTQ_DIR", help="Folder containing raw FastQ files.")
+    parser.add_argument("SAMPLESHEET_FILE", help="Output samplesheet file.")
+    parser.add_argument('-st', '--strandedness', type=str, dest="STRANDEDNESS", default='unstranded', help="Value for 'strandedness' in samplesheet. Must be one of 'unstranded', 'forward', 'reverse'.")
+    parser.add_argument('-r1', '--read1_extension', type=str, dest="READ1_EXTENSION", default='_R1_001.fastq.gz', help="File extension for read 1.")
```
How about specifying a glob pattern instead of an extension? I don't see any downsides and it would be more flexible.

drpatelh replied:
Can't visualise how this would look in the code. Fancy coming up with a small PR adding this after this has been merged?

grst commented on bin/fastq_dir_to_samplesheet.py:
```diff
@@ -0,0 +1,87 @@
+        sample = os.path.basename(read1_file).replace(read1_extension,'')
+        if sanitise_name:
+            sample = sanitise_name_delimiter.join(os.path.basename(read1_file).split(sanitise_name_delimiter)[:sanitise_name_index])
+        read_dict[sample] = [read1_file]
```
Does this still work if samples have been re-sequenced (to be `cat`ed by the pipeline)?

drpatelh replied:
Yes, it should do but you would need to use the sanitise parameters in the script to get just the main id of the pipeline and strip out all of the lane info etc.

grst replied:
Hang on, but the dict key is the sample_id, so there can only be one pair of fastq files per sample id. Resequenced samples would require two lines in the samplesheet with the same sample_id. Am I missing something?

drpatelh replied:
You are absolutely right 🤦🏽 I initially wrote this script for viralrecon where merging isn't really required but I should have tested a little more thoroughly. 

Fixed https://github.com/nf-core/rnaseq/pull/660/commits/17a15fe233e3f9d6a53744dc4ed45b43a910cdb9

Given the directory structure below:

```console
$ mkdir fastq
$ cd fastq
$ touch DAN2501A41_S3_L001_R1_001.fastq.gz
$ touch DAN2501A41_S3_L001_R2_001.fastq.gz
$ touch DAN2501A41_S3_L002_R1_001.fastq.gz
$ touch DAN2501A41_S3_L002_R2_001.fastq.gz
$ touch DAN2501A42_S4_L001_R1_001.fastq.gz
$ touch DAN2501A42_S4_L001_R2_001.fastq.gz
$ touch DAN2501A42_S4_L002_R1_001.fastq.gz
$ touch DAN2501A42_S4_L002_R2_001.fastq.gz
$ touch DAN2501A43_S5_L001_R1_001.fastq.gz
$ touch DAN2501A43_S5_L001_R2_001.fastq.gz
```

Running the command below will add separate entries for each library:

```console
./fastq_dir_to_samplesheet.py fastq test.csv --sanitise_name
```

grst replied:
LGTM now!

grst commented on bin/sra_ids_to_runinfo.py:
```diff
@@ -8,14 +8,26 @@
 
-
 ## Example ids supported by this script
 SRA_IDS = ['PRJNA63463', 'SAMN00765663', 'SRA023522', 'SRP003255', 'SRR390278', 'SRS282569', 'SRX111814']
```
I though handling of external samples has moved completely to the `fetchngs` pipeline? Shouldn't this script move there as well?

drpatelh replied:
Fixed https://github.com/nf-core/rnaseq/pull/660/commits/57f3a4760ebd78fd4252458f3f8088d6bef5c2c6

grst commented on bin/sra_runinfo_to_ftp.py:
```diff
@@ -6,7 +6,6 @@
 import argparse
 import collections
```
same for this script...

drpatelh replied:
Fixed https://github.com/nf-core/rnaseq/pull/660/commits/57f3a4760ebd78fd4252458f3f8088d6bef5c2c6

maxulysse reviewed (commented):
Most of @grst comments are on point, and I'm afraid I have very little to add.
I agree that the removing of SRA download could be seen as a major upgrade, but I think this was a minor feature so I think it's still ok to go for a 3.2 release.

JoseEspinosa commented on main.nf:
```diff
@@ -40,22 +40,13 @@ WorkflowMain.initialise(workflow, params, log)
-    if (params.public_data_ids) {
-        include { SRA_DOWNLOAD } from './workflows/sra_download'
-        SRA_DOWNLOAD ()
+include { RNASEQ } from './workflows/rnaseq'
```
It became a bug now see [here](https://github.com/nextflow-io/nextflow/commit/7dbf64bea38907126f44b09a023b8061bf3363d0) :open_mouth:

drpatelh replied:
Yup. Saw that! Fixed, just in the nick of time 😆

JoseEspinosa reviewed (approved):
LGTM! 
Ready to go after addressing @grst suggestions. :rocket:

grst commented on bin/fastq_dir_to_samplesheet.py:
```diff
@@ -0,0 +1,87 @@
+    parser.add_argument("FASTQ_DIR", help="Folder containing raw FastQ files.")
+    parser.add_argument("SAMPLESHEET_FILE", help="Output samplesheet file.")
+    parser.add_argument('-st', '--strandedness', type=str, dest="STRANDEDNESS", default='unstranded', help="Value for 'strandedness' in samplesheet. Must be one of 'unstranded', 'forward', 'reverse'.")
+    parser.add_argument('-r1', '--read1_extension', type=str, dest="READ1_EXTENSION", default='_R1_001.fastq.gz', help="File extension for read 1.")
+    parser.add_argument('-r2', '--read2_extension', type=str, dest="READ2_EXTENSION", default='_R2_001.fastq.gz', help="File extension for read 2.")
+    parser.add_argument('-se', '--single_end', dest="SINGLE_END", action='store_true', help="Single-end information will be auto-detected but this option forces paired-end FastQ files to be treated as single-end so only read 1 information is included in the samplesheet.")
+    parser.add_argument('-sn', '--sanitise_name', dest="SANITISE_NAME", action='store_true', help="Whether to further sanitise FastQ file name to get sample id. Used in conjunction with --sanitise_name_delimiter and --sanitise_name_index.")
+    parser.add_argument('-sd', '--sanitise_name_delimiter', type=str, dest="SANITISE_NAME_DELIMITER", default='_', help="Delimiter to use to sanitise sample name.")
+    parser.add_argument('-si', '--sanitise_name_index', type=int, dest="SANITISE_NAME_INDEX", default=1, help="After splitting FastQ file name by --sanitise_name_delimiter all elements before this index (1-based) will be joined to create final sample name.")
+    return parser.parse_args(args)
+
+def fastq_dir_to_samplesheet(fastq_dir, samplesheet_file, strandedness='unstranded', read1_extension='_R1_001.fastq.gz', read2_extension='_R2_001.fastq.gz', single_end=False, sanitise_name=False, sanitise_name_delimiter='_', sanitise_name_index=1):
+
+    ## Get read 1 files
+    read_dict = {}
+    read1_files = glob.glob(os.path.join(fastq_dir, f'*{read1_extension}'), recursive=False)
```
```suggestion
    parser.add_argument('-r1', '--read1_pattern', type=str, dest="READ1_EXTENSION", default='*_R1_001.fastq.gz', help="File extension for read 1.")
    parser.add_argument('-r2', '--read2_pattern', type=str, dest="READ2_EXTENSION", default='*_R2_001.fastq.gz', help="File extension for read 2.")
    parser.add_argument('-se', '--single_end', dest="SINGLE_END", action='store_true', help="Single-end information will be auto-detected but this option forces paired-end FastQ files to be treated as single-end so only read 1 information is included in the samplesheet.")
    parser.add_argument('-sn', '--sanitise_name', dest="SANITISE_NAME", action='store_true', help="Whether to further sanitise FastQ file name to get sample id. Used in conjunction with --sanitise_name_delimiter and --sanitise_name_index.")
    parser.add_argument('-sd', '--sanitise_name_delimiter', type=str, dest="SANITISE_NAME_DELIMITER", default='_', help="Delimiter to use to sanitise sample name.")
    parser.add_argument('-si', '--sanitise_name_index', type=int, dest="SANITISE_NAME_INDEX", default=1, help="After splitting FastQ file name by --sanitise_name_delimiter all elements before this index (1-based) will be joined to create final sample name.")
    return parser.parse_args(args)

def fastq_dir_to_samplesheet(fastq_dir, samplesheet_file, strandedness='unstranded', read1_extension='_R1_001.fastq.gz', read2_extension='_R2_001.fastq.gz', single_end=False, sanitise_name=False, sanitise_name_delimiter='_', sanitise_name_index=1):

    ## Get read 1 files
    read_dict = {}
    read1_files = glob.glob(os.path.join(fastq_dir, f'{read1_extension}'), recursive=False)
```

grst replied:
@drpatelh, actually, this was all I meant ;) 
(+ maybe updating the descriptions) 

But we can do that later if it's getting too much for this PR.

drpatelh replied:
Cool. Yep. Be nice to do in another PR if poss please :)

drpatelh commented:
Think I have addressed all of your comments now @grst! Thanks for the thorough review :)

Be awesome if we can create a small PR adding the regex functionality and then we can release.

drpatelh commented:
We good to go now @grst ? Be awesome if you can approved. Sorry, tending to a sick baby so be good to get this out before **** hits the fan again!

grst commented:
Let's go! :boom:
